Add support for custom annotations in the CSS BNF syntax

weinig opened the pull request:
#### 78551d0c1acbf2e69884ad14f993d45f4ca1caed
<pre>
Add support for custom annotations in the CSS BNF syntax
<a href="https://bugs.webkit.org/show_bug.cgi?id=250010">https://bugs.webkit.org/show_bug.cgi?id=250010</a>
rdar://103817055

Reviewed by Simon Fraser.

Adds suppport for a custom syntax to annotate the CSS property grammar
BNF to allow passing additional information to the generators on a per
term basis. The syntax is of the form `@(foo-bar baz)`. That is an &apos;@&apos;
symbol followed by a space separated list of identifiers inside a pair
of parentheses.

In this initial commit, it is used to annotate a &apos;#&apos; multiplier as:

  &quot;normal | &lt;variation-tag-value&gt;#@(no-single-item-opt)&quot;

This will tell the code generators to not use the default behavior of
stashing a single item list as the item itself, but rather to always
return a list.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontVariationTag): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontVariationSettings): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontTag):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFeatureTagValue):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFeatureSettings):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeVariationTagValue):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFeatureTag): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h:
* Source/WebCore/css/process-css-properties.py:

Canonical link: <a href="https://commits.webkit.org/258410@main">https://commits.webkit.org/258410@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d89f50f34e0891b5d67cc9f5aaee18fdd8ed674

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101873 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11022 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34947 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111208 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1936 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/108964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107653 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/4608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4694 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1788 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/10773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44837 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5767 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6446 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -522,27 +523,44 @@ RefPtr<CSSValueList> consumeFontFaceUnicodeRange(CSSParserTokenRange& range)
+
+    // FIXME: The specification states "The <string> is a case-sensitive OpenType feature tag."
+    // so we probably should not be lowercasing it at parse time.
+    auto tag = consumeFontTag<true>(range);
```
Hard to know what the `true` means

weinig replied:
I'll replace with an enum.

Commit:
Add support for custom annotations in the CSS BNF syntax
https://bugs.webkit.org/show_bug.cgi?id=250010
rdar://103817055

Reviewed by Simon Fraser.

Adds suppport for a custom syntax to annotate the CSS property grammar
BNF to allow passing additional information to the generators on a per
term basis. The syntax is of the form `@(foo-bar baz)`. That is an '@'
symbol followed by a space separated list of identifiers inside a pair
of parentheses.

In this initial commit, it is used to annotate a '#' multiplier as:

  "normal | <variation-tag-value>#@(no-single-item-opt)"

This will tell the code generators to not use the default behavior of
stashing a single item list as the item itself, but rather to always
return a list.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontVariationTag): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontVariationSettings): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontTag):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFeatureTagValue):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFeatureSettings):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeVariationTagValue):
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFeatureTag): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h:
* Source/WebCore/css/process-css-properties.py:

Canonical link: https://commits.webkit.org/258410@main
